docs/bookstore/page-2/sentiment-analysis-service-for-bookstore-reviews.md

```diff
@@ -0,0 +1,414 @@
+func deploy -b=s2i -v
+```
+???+ success "Verify"
+
+    When the deployment is complete, you will see the following output:
+    
+    ```
+    Function deployed in namespace "default" and exposed at URL:
+    http://sentiment-analysis-app.default.svc.cluster.local
+    ```
+ 
+!!! tip  
+    You can find the URL of the Knative Function (Knative Service) by running the following command:
+    
+    ```bash
+    kubectl get kservice
+    ```
+    
+    You will see the URL in the output:
+    
+    ```
+    NAME                     URL                                                       LATESTCREATED                  LATESTREADY                    READY   REASON
+    sentiment-analysis-app   http://sentiment-analysis-app.default.svc.cluster.local   sentiment-analysis-app-00001   sentiment-analysis-app-00001   True    
+    ```
+
+
+
+## **Knative Serving: scale down to zero**
+
+![Image13](images/image13.png)
+
+If you use the following command to query all the pods in the cluster, you will see that the pod is running:
+
+```bash
+kubectl get pods
+```
+
+where `-A` is the flag to query all the pods in all namespaces.
+
+And you will find that your sentiment analysis app is running:
+
+```
+NAMESPACE   NAME                                      READY   STATUS    RESTARTS   AGE
+default     sentiment-analysis-app-00002-deployment   2/2     Running   0          2m
+```
+
+But if you wait for a while without sending any CloudEvent to your function, and query the pods again, you will find that the pod that has your sentiment analysis app **disappeared**!
+
+This is because Knative Serving's autoscaler will **automatically scale down to zero** if there is no request to the function! Learn more at [Knative Autoscaling](https://knative.dev/docs/serving/autoscaling/){:target="_blank"}.
+
+---
+
+## **Verify**
+
+![Image2](images/image2.png)
+
+After deployment, the `func` CLI provides a URL to access your function. You can verify the function's operation by sending a request with a sample review comment.
+
+Simply use Knative Function's command `func invoke` to directly send a CloudEvent to the function on your cluster:
+
+```bash
+func invoke -f=cloudevent --data='{"reviewText":"I love Knative so much"}' -v
+```
+
+- `-f` flag indicates the type of the data, is either `HTTP` or `cloudevent`
+- `--data` flag is the input text
+- `-t` flag is the URI to the Knative Function.
+
+???+ success "Verify"
+
+    If you see the response, it means that the function is running successfully.
+
+    ```
+    Context Attributes,
+      specversion: 1.0
+      type: moderated-comment
+      source: sentiment-analysis
+      id: 0c2d0659-a30e-4efd-bcce-803f15ff5cc5
+      time: 2024-06-11T15:12:43.795405Z
+      datacontenttype: application/json
+    Extensions,
+      badwordfilter: 
+      sentimentresult: positive
+    Data,
+      {
+        "reviewText": "I love Knative so much",
+        "badWordResult": "",
+        "sentimentResult": "positive"
+      }
```

Comment:
I don't see a response
```
pierdipi@pierdipi ML-sentiment-analysis (compile-pages) $ func deploy -b=s2i -v --registry quay.io/pierdipi --build
Warning: function has current image 'quay.io/rh-ee-leoli/sentiment-analysis-app:latest' which has a different registry than the currently configured registry 'quay.io/pierdipi'. The new image tag will be 'quay.io/pierdipi/sentiment-analysis-app:latest'.  To use an explicit image, use --image.
Building function image
Step 1/8 : FROM registry.access.redhat.com/ubi8/python-39
latest: Pulling from ubi8/python-39
Digest: sha256:78f4d1a6528fde88dd8aaee321d3fadceccb52ad1c8781eb13370a5933bd1ba0
Status: Image is up to date for registry.access.redhat.com/ubi8/python-39:latest
 ---> 07d8b13135a7
Step 2/8 : LABEL "io.k8s.display-name"="quay.io/pierdipi/sentiment-analysis-app:latest"       "io.openshift.s2i.build.image"="registry.access.redhat.com/ubi8/python-39"       "io.openshift.s2i.build.source-location"="/home/pierdipi/redhat/knative/docs-1/code-samples/eventing/bookstore-sample-app/solution/ML-sentiment-analysis"       "io.openshift.s2i.scripts-url"="image:///usr/libexec/s2i"
 ---> Using cache
 ---> 139c1b410b10
Step 3/8 : USER root
 ---> Using cache
 ---> 9acce44840b1
Step 4/8 : COPY upload/src /tmp/src
 ---> Using cache
 ---> dc5659dfcd61
Step 5/8 : RUN chown -R 1001:0 /tmp/src
 ---> Using cache
 ---> a01e70f12089
Step 6/8 : USER 1001
 ---> Using cache
 ---> 588c0487fbf9
Step 7/8 : RUN /usr/libexec/s2i/assemble
 ---> Using cache
 ---> 64c2c6bc7a1a
Step 8/8 : CMD /usr/libexec/s2i/run
 ---> Using cache
 ---> ac2b63129513
Successfully built ac2b63129513
Successfully tagged quay.io/pierdipi/sentiment-analysis-app:latest
🙌 Function built: quay.io/pierdipi/sentiment-analysis-app:latest
Pushing function image to the registry "quay.io" using the "pierdipi" user credentials
The push refers to repository [quay.io/pierdipi/sentiment-analysis-app]
81bf7fcae605: Preparing
00f5dab2e5a8: Preparing
06fa391137b2: Preparing
7e62c82dccdd: Preparing
3510106f4711: Preparing
acf02a0b1572: Preparing
75401ea2a612: Preparing
acf02a0b1572: Waiting
75401ea2a612: Waiting
3510106f4711: Layer already exists
81bf7fcae605: Layer already exists
06fa391137b2: Layer already exists
75401ea2a612: Layer already exists
acf02a0b1572: Layer already exists
00f5dab2e5a8: Layer already exists
7e62c82dccdd: Layer already exists
latest: digest: sha256:b9d09ba701aec277c092feff03494b4126878860e68b8aefa11bd4a166b8b441 size: 1794
⬆️  Deploying function to the cluster
---> Running application from script (app.sh) ...
---> Running application from script (app.sh) ...
✅ Function updated in namespace "default" and exposed at URL: 
   http://sentiment-analysis-app.default.127.0.0.1.sslip.io
pierdipi@pierdipi ML-sentiment-analysis (compile-pages) $ func invoke -f=cloudevent --data='{"reviewText":"I love Knative so much"}' -v
Invoking '' function at http://sentiment-analysis-app.default.127.0.0.1.sslip.io
Invoking '' function using 'cloudevent' format
Sending event
Context Attributes,
  specversion: 1.0
  type: boson.fn
  source: /boson/fn
  id: 
  datacontenttype: application/json
Data,
  "{\"reviewText\":\"I love Knative so much\"}"
Function invoked.  Response:
pierdipi@pierdipi ML-sentiment-analysis (compile-pages) $ k get pods 
NAME                                                       READY   STATUS    RESTARTS   AGE
bookstore-frontend-7b879ffb78-nrp24                        1/1     Running   0          45m
event-display-55967c745d-77f5c                             1/1     Running   0          41m
node-server-644795d698-8bwtj                               1/1     Running   0          42m
sentiment-analysis-app-00001-deployment-779fffbdd7-9fhqw   2/2     Running   0          118s
sentiment-analysis-app-00002-deployment-78d49b6795-7272j   2/2     Running   0          55s
```

Reply:
Also it's weird `id` is empty after `Sending event`

cc @lkingland @matejvasek

Reply:
I just tested it, it works fine on my end. The response is there, and the `id` is non-empty.

```
func invoke -f=cloudevent --data='{"reviewText":"I love Knative so much"}' -v                              ✔  1m 55s  16:52:33 
Invoking '' function at http://sentiment-analysis-app.default.svc.cluster.local
Invoking '' function using 'cloudevent' format
Sending event
Context Attributes,
  specversion: 1.0
  type: boson.fn
  source: /boson/fn
  id: 
  datacontenttype: application/json
Data,
  {
    "reviewText": "I love Knative so much"
  }

Function invoked.  Response:
Context Attributes,
  specversion: 1.0
  type: moderated-comment
  source: sentiment-analysis
  id: de3d9c80-1e40-481a-8cfa-3c17c3adf6d7
  time: 2024-06-20T20:54:10.208172Z
  datacontenttype: application/json
Extensions,
  badwordfilter: 
  sentimentresult: positive
Data,
  {
    "reviewText": "I love Knative so much",
    "badWordResult": "",
    "sentimentResult": "positive"
  }
```

Maybe try the new `func.yaml` file [here](https://github.com/Leo6Leo/docs/blob/fix-comments/code-samples/eventing/bookstore-sample-app/solution/ML-sentiment-analysis/func.yaml#L1)?

Reply:
I am also seeing a response:

```
func invoke -f=cloudevent --data='{"reviewText":"I love Knative so much"}' -v

Invoking '' function at http://sentiment-analysis-app.default.10.99.189.63.sslip.io
Invoking '' function using 'cloudevent' format
Sending event
Context Attributes,
  specversion: 1.0
  type: boson.fn
  source: /boson/fn
  id: 
  datacontenttype: application/json
Data,
  {
    "reviewText": "I love Knative so much"
  }
Function invoked.  Response:
Context Attributes,
  specversion: 1.0
  type: moderated-comment
  source: sentiment-analysis
  id: 10a746c8-dd35-433c-9497-e6efcb5f946b
  time: 2024-06-25T12:07:12.923682Z
  datacontenttype: application/json
Extensions,
  badwordfilter: 
  sentimentresult: positive
Data,
  {
    "reviewText": "I love Knative so much",
    "badWordResult": "",
    "sentimentResult": "positive"
  }
```